sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -95,13 +95,12 @@ private[sql] object Dataset {
   def ofRows(sparkSession: SparkSession, logicalPlan: LogicalPlan): DataFrame =
     sparkSession.withActive {
       val qe = sparkSession.sessionState.executePlan(logicalPlan)
-      val encoder = if (qe.isLazyAnalysis) {
-        RowEncoder.encoderFor(new StructType())
+      if (qe.isLazyAnalysis) {
```

Comment:
can we clean up the code a bit more?
```
if (!qe.isLazyAnalysis) qe.assertAnalyzed()
new Dataset[Row](qe, () => RowEncoder.encoderFor(qe.analyzed.schema))
```

Reply:
Now with this PR, the encoder creation is lazy only when `qe.isLazyAnalysis`, to be consistent with the current behavior. Also the previously failed test suggested to make it consistent.

Reply:
I don't get it. My proposal is a pure code cleanup that doesn't change any actual logic.

Reply:
The lazy encoder creation only happens when `qe.isLazyAnalysis`; otherwise create it right here as same as before to avoid changing the behavior. `Dataset` has a new default constructor to take a function for the lazy case, and the original default constructor is still there to keep the behavior.

Reply:
I get it that we want to create RowEncoder eagerly to fail earlier for the common case, but `if (!qe.isLazyAnalysis) qe.assertAnalyzed()` should do the same thing, right?

Reply:
Sure, I updated.

I'm not sure [this change](https://github.com/apache/spark/pull/48829/files#diff-0d94d471b64499a7962de4e9c3335ff5227b80cd85e50cfbd4cab67e41651fb8R1208) is allowed. I guess we still want to create it eagerly if it's not lazy analysis?